### PR TITLE
Factor out shorthand serialization into its own class

### PR DESCRIPTION
**Pull request by darinadler**
#### 66b28324b8b97037caca3b9f1c9d2f724abaaf7d
<pre>
Factor out shorthand serialization into its own class
<a href="https://bugs.webkit.org/show_bug.cgi?id=250953">https://bugs.webkit.org/show_bug.cgi?id=250953</a>
rdar://problem/104517006

Reviewed by Oriol Brufau.

Factoring out shorthand serialization into a separate class and breaking it out of StyleProperties
has a few benefits:

- Moves details of shorthand serialization out of StyleProperties.h; these are no longer private
members of StyleProperties class.
- Reorganizes shorthand serialization so longhand property values are gathered during the common
checks. This means we won&apos;t repeatedly search style properties for the same properties.
- Prepares for moving shorthand serialization into a separate file, making StyleProperties.cpp
a more manageable size.
- Prepares for sharing shorthand serialization with computed style. We don&apos;t need to repeat all
this logic twice the way we do today. However, computed style has a WebKit-only feature where it
will produce CSS values for shorthands; we have to decide whether to remove that entirely or at
least remove if for most shorthand properties. Shorthands don&apos;t need to be expressible as CSS
values; they are a parsing and serialization format and don&apos;t have a specified object model.

Also made a small number of other minor improvements to StyleProperties.

* Source/WebCore/css/CSSProperties.json: Reordered the longhands in border, border-block, and
border-inline so they are in serialization order: width, style, color. The new serialization code
depends on this, and it&apos;s also a logical order.

* Source/WebCore/css/StyleProperties.cpp: Updated the maxShorthandLength constant, which was
incorrect. It&apos;s still a good idea to generate it and we should do that in a future change.
(WebCore::isCSSWideValueKeyword): Deleted.
(WebCore::isValueIDIncludingList): Removed an unused overload.
(WebCore::serializeLonghandValue): Renamed from textExpandingInitialValuePlaceholder and moved
the special logic for opacity properties in here. Also added an overload that takes a pointer.
(WebCore::ShorthandSerializer::commonSerializationChecks): Renamed from
SyleProperties::commonShorthandChecks. Assert length is not zero since ShorthandSerializer must
only be used for shorthands. Use a boolean return value instead of using empty string and null
string to express the two different outcomes. Store all the values in m_longhandValues as we check
them. Also added the m_gridTemplateAreasWasSetFromShorthand logic.
(WebCore::StyleProperties::serializeLonghandValue const): Added. Uses serializeLonghandValue and
getPropertyCSSValue.
(WebCore::StyleProperties::serializeShorthandValue const): Added. Uses ShorthandSerializer.
(WebCore::StyleProperties::getPropertyValue const): Use serializeLonghandValue and
serializeShorthandValue.
(WebCore::ShorthandSerializer::ShorthandSerializer): Added.
(WebCore::ShorthandSerializer::longhandProperty const): Added.
(WebCore::ShorthandSerializer::longhandValue const): Added.
(WebCore::ShorthandSerializer::serializeValue): Added.
(WebCore::ShorthandSerializer::isInitialValue): Added.
(WebCore::ShorthandSerializer::longhandIndex const): Added.
(WebCore::ShorthandSerializer::longhandValueID const): Added.
(WebCore::ShorthandSerializer::serializeLonghandValue const): Added.
(WebCore::ShorthandSerializer::serialize): Added. Moved the large switch statment from
StyleProperties::getPropertyValue in here. Besides name changes, changed serialization specifics:
for &quot;all&quot;, don&apos;t try to find a common value, since the only ones are the CSS-wide keywords already
handled by the common serialization checks. For border, border-block, and border-inline, pass in
the size of each section of the shorthand, allowing us to simplify the shared code. For container,
use serializeLonghandsOmittingTrailingInitialValue. For flex and perspective-origin, use
serializeLonghands. for transform-origin, use serializeLonghandsOmittingTrailingInitialValue.
(WebCore::StyleProperties::propertyAsColor const): Use serializeLonghandValue.
(WebCore::ShorthandSerializer::serializeFont const): Get longhand values by index. Use makeString.
(WebCore::ShorthandSerializer::serializeOffset const): Ditto.
(WebCore::ShorthandSerializer::serializeFontVariant const): Use
serializeLonghandsOmittingInitialValues after doing some additional needed checks.
(WebCore::ShorthandSerializer::serializeFontSynthesis const): Get longhand values by index.
Use ASCII literals instead of StringBuilder.
(WebCore::ShorthandSerializer::serializePair const): Get longhand values by index.
(WebCore::ShorthandSerializer::serializeQuad const): Ditto.
(WebCore::LayerValues::serialize const): Use serializeLonghandValue.
(WebCore::ShorthandSerializer::serializeLayered const): Get longhand values by index.
(WebCore::ShorthandSerializer::serializeGridTemplate const): Ditto.
(WebCore::gridAutoFlowContains): Argument type is now CSSValue&amp;.
(WebCore::ShorthandSerializer::serializeGrid const): Get longhand values by index. Use makeString.
(WebCore::ShorthandSerializer::serializeGridRowColumn const): Use serializeLonghands.
(WebCore::ShorthandSerializer::serializeGridArea const): Ditto.
(WebCore::ShorthandSerializer::serializeLonghands const): Added. This is like
StyleProperties::getShorthandValue was, but does not omit initial values and is more versatile
since it can be used to serialize only some properties up to a limit. Also use makeString for
common short lengths.
(WebCore::ShorthandSerializer::serializeLonghandsOmittingInitialValues const): This is the new
name for StyleProperties::getShorthandValue.
(WebCore::ShorthandSerializer::serializeLonghandsOmittingTrailingInitialValue const): Added.
Convenient for cases where the last value is optional, but other values are not.
(WebCore::ShorthandSerializer::serializeCommonValue const): Removed unneeded code to handle
missing longhands; that&apos;s checked by commonSerializationChecks. Refactored into two functions,
so this can be used on a subset of the shorthand or the whole thing.
(WebCore::ShorthandSerializer::serializeBorderImage const): Get longhand values by index.
(WebCore::ShorthandSerializer::serializeBorderRadius const): Ditto.
(WebCore::ShorthandSerializer::serializeBorder const): Use a new approach where we serialize
these border-related shorthands by section. Get longhand values by index. Use
serializeCommonValue and subsequentLonghandsHaveInitialValues.
(WebCore::ShorthandSerializer::serializeBreakInside const): Use longhandValueID.
(WebCore::ShorthandSerializer::serializePageBreak const): Ditto.
(WebCore::ShorthandSerializer::serializeColumnBreak const): Ditto.
(WebCore::ShorthandSerializer::subsequentLonghandsHaveInitialValues const): Replaces
hasAllInitialValues.
(WebCore::MutableStyleProperties::removeShorthandProperty): Added a returnText argument so we can
implement the return value for shorthands here. Still using a FIXME for this for now, we&apos;ll make
that change separately.
(WebCore::MutableStyleProperties::removePropertyAtIndex): Added. Helps factor out common
code and give us slightly better efficiency. Also use null string to mean no result rather
than empty string.
(WebCore::MutableStyleProperties::removeLonghandProperty): Added. Calls findPropertyIndex and
removePropertyAtIndex.
(WebCore::MutableStyleProperties::removeProperty): Call removeShorthandProperty and
removeLonghandProperty.
(WebCore::MutableStyleProperties::removeCustomProperty): Call removePropertyAtIndex.
(WebCore::MutableStyleProperties::setProperty): Removed the unneeded overload for the case where
the property&apos;s value is a property ID. Removed the unneeded call to removeShorthandProperty
in the overload that takes a CSSProperty that is never called for shorthands, adding an
assertion to check that invariant.
(WebCore::StyleProperties::asTextInternal const): Use serializeShorthandValue and
serializeLonghandValue.
(WebCore::MutableStyleProperties::removePropertiesInSet): Use the HashSet::add function that
takes iterators.
(WebCore::ImmutableStyleProperties::findCustomPropertyIndex const): Take a StringView instead
of a const String&amp;.
(WebCore::MutableStyleProperties::findCustomPropertyIndex const): Ditto.
(WebCore::MutableStyleProperties::findCSSPropertyWithID): Use nullptr.
(WebCore::MutableStyleProperties::findCustomCSSPropertyWithName): Ditto.
(WebCore::StyleProperties::PropertyReference::cssText const): Use serializeLonghandValue.

* Source/WebCore/css/StyleProperties.h: Made findPropertyIndex and findCustomPropertyIndex
public in StyleProperties. There are other functions that work based on the index in the
interface. Doing this allows ShorthandSerializer to work using only public functions.
Initialize m_arraySize to 0. Removed unneeded default constructor, most private functions,
and friend relationship with PropertySetCSSStyleDeclaration. Updated MutableStyleProperties
functions for changes above.

* Source/WebCore/css/StylePropertyShorthand.h: Removed unneeded
fontShorthandSubpropertiesResetToInitialValues; we now base this on CSSProperties.json.
Added functions mentioned above.

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeFont): Set to initial values intead of explicitly
allocating CSSValueNormal. This may be slightly more efficient and it&apos;s consistent with
other property parsing. Set all properties in a loop, takes care of the properties
reset to initial values without fontShorthandSubpropertiesResetToInitialValues.

Canonical link: <a href="https://commits.webkit.org/259211@main">https://commits.webkit.org/259211@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/7e0b7a5a2fe5b89b74352142a0e3e30915d5f617

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104347 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13433 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37259 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113563 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173856 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108272 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14526 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4347 "Built successfully") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96572 "Failed to checkout and rebase branch from PR 8956") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112604 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110115 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11195 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94255 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/96572 "Failed to checkout and rebase branch from PR 8956") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93058 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25866 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/96572 "Failed to checkout and rebase branch from PR 8956") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6793 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27223 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6922 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3777 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12944 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46781 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6351 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8712 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebCore/css/StyleProperties.cpp [hunk 1]
```diff
@@ -214,45 +317,83 @@ String StyleProperties::commonShorthandChecks(const StylePropertyShorthand& shor
...
-            return textExpandingInitialValuePlaceholder(propertyID, *value);
-        }
-    }
+    if (auto value = getPropertyCSSValue(propertyID))
+        return serializeLonghandValue(propertyID, *value);
 
-    auto shorthand = shorthandForProperty(propertyID);
-    if (!shorthand.length())
-        return String();
+    return ShorthandSerializer(*this, propertyID).serialize();
```
Comment (Loirooriol) [inline Source/WebCore/css/StyleProperties.cpp:348]: It seems a bit strange to use something called `ShorthandSerializer` for a longhand. And it was already the case, but using `getPropertyCSSValue` for shorthands was futile. Maybe

```cpp
if (!isLonghand(propertyID))
    return ShorthandSerializer(*this, propertyID).serialize();
if (auto value = getPropertyCSSValue(propertyID))
    return serializeLonghandValue(propertyID, *value);
return String();
```

And then in `ShorthandSerializer::commonSerializationChecks` assert `length() > 0`.

File: Source/WebCore/css/parser/CSSPropertyParser.cpp [hunk 2]
```diff
@@ -667,27 +667,24 @@ bool CSSPropertyParser::consumeFont(bool important)
...
     add(CSSPropertyFontSize, fontSize);
     add(CSSPropertyLineHeight, lineHeight);
     add(CSSPropertyFontFamily, fontFamily);
-    for (auto [property, initialValue] : fontShorthandSubpropertiesResetToInitialValues)
-        reset(property, initialValue);
+    bool pastFontFamily = false;
+    for (auto longhand : fontShorthand()) {
+        if (longhand == CSSPropertyFontFamily)
+            pastFontFamily = true;
+        else if (pastFontFamily)
```
Comment (Loirooriol) [inline Source/WebCore/css/parser/CSSPropertyParser.cpp]: This is already assuming that `font-family` is the 7th longhand, so why not just start the loop there?

File: Source/WebCore/css/StyleProperties.cpp [hunk 3]
```diff
@@ -1225,58 +1317,44 @@ RefPtr<CSSValue> StyleProperties::getCustomPropertyCSSValue(const String& proper
...
-bool MutableStyleProperties::removeShorthandProperty(CSSPropertyID propertyID)
+bool MutableStyleProperties::removeShorthandProperty(CSSPropertyID propertyID, String* returnText)
 {
-    StylePropertyShorthand shorthand = shorthandForProperty(propertyID);
+    auto shorthand = shorthandForProperty(propertyID);
     if (!shorthand.length())
         return false;
-
+    if (returnText)
+        *returnText = getPropertyValue(propertyID);
```
Comment (Loirooriol) [inline Source/WebCore/css/StyleProperties.cpp]: `removeProperty` was just setting `*returnText = emptyString()` in case of shorthands.

So this seems to be a webexposed unrelated fix that should probably have its own patch, and be tested.

File: Source/WebCore/css/StyleProperties.cpp [hunk 4]
```diff
@@ -1603,7 +1676,7 @@ StringBuilder StyleProperties::asTextInternal() const
             continue;
 
         if (value.isNull())
-            value = textExpandingInitialValuePlaceholder(propertyID, *property.value());
+            value = serializeLonghandValue(propertyID, *property.value());
```
Comment (Loirooriol) [inline Source/WebCore/css/StyleProperties.cpp]: Nit: to be consistent with this being explicit about serializing a longhand, the loop above could use `ShorthandSerializer(*this, shorthandPropertyID).serialize()` (or maybe add `serializeShorthandValue` as an alias?)

File: Source/WebCore/css/StyleProperties.cpp [hunk 5]
```diff
@@ -1225,58 +1317,44 @@ RefPtr<CSSValue> StyleProperties::getCustomPropertyCSSValue(const String& proper
...
-    }
-
-    if (returnText)
-        *returnText = propertyAt(foundPropertyIndex).value()->cssText();
-
-    // A more efficient removal strategy would involve marking entries as empty
-    // and sweeping them when the vector grows too big.
-    m_propertyVector.remove(foundPropertyIndex);
+    return removeShorthandProperty(propertyID, returnText)
+        || removePropertyAtIndex(findPropertyIndex(propertyID), returnText);
```
Comment (Loirooriol) [inline Source/WebCore/css/StyleProperties.cpp]: If `propertyID` is a shorthand but no longhand is present, no need to search again.

```cpp
if (isLonghand(propertyID))
    return removePropertyAtIndex(findPropertyIndex(propertyID), returnText);
return removeShorthandProperty(propertyID, returnText);
```

File: Source/WebCore/css/StyleProperties.cpp [hunk 6]
```diff
@@ -1118,15 +1202,33 @@ String StyleProperties::borderRadiusShorthandValue(const StylePropertyShorthand&
...
+    };
+    auto widthText = serializeSection(0);
     if (widthText.isNull())
         return String();
-    auto styleText = getCommonValue(style);
+    auto styleText = serializeSection(sectionLength);
     if (styleText.isNull())
         return String();
-    auto colorText = getCommonValue(color);
+    auto colorText = serializeSection(sectionLength + sectionLength);
```
Comment (Loirooriol) [inline Source/WebCore/css/StyleProperties.cpp]: Nit: you are using `* 3` above, so this could use `* 2`.

File: Source/WebCore/css/StyleProperties.cpp [hunk 7]
```diff
@@ -967,64 +1046,73 @@ static bool canOmitTrailingGridAreaValue(CSSValue& value, CSSValue& trailing)
...
-String StyleProperties::getGridRowColumnShorthandValue(const StylePropertyShorthand& shorthand) const
+String ShorthandSerializer::serializeGridRowColumn() const
 {
-    ASSERT(shorthand.length() == 2);
-    auto start = getPropertyCSSValue(shorthand.properties()[0]);
-    auto end = getPropertyCSSValue(shorthand.properties()[1]);
-    if (canOmitTrailingGridAreaValue(*start, *end))
-        return start->cssText();
-    return makeString(start->cssText(), " / ", end->cssText());
+    return serializeLonghands(canOmitTrailingGridAreaValue(longhandValue(0), longhandValue(1)) ? 1 : 2, " / ");
```
Comment (Loirooriol) [inline Source/WebCore/css/StyleProperties.cpp:1081]: I think I would keep the assert, in case a new longhand is added in the future.
Ditto for `serializeGridArea`.

File: Source/WebCore/css/StyleProperties.cpp [hunk 8]
```diff
@@ -898,61 +988,50 @@ static bool gridAutoFlowContains(const RefPtr<CSSValue>& autoFlow, CSSValueID id
...
-        result.append(autoRows->cssText());
-    }
-
-    result.append(" / ");
-    result.append(columns->cssText());
-
-    return result.toString();
+    if (isValueIDIncludingList(autoRows, CSSValueAuto))
+        return makeString("auto-flow", dense, " / ", serializeLonghandValue(columnsIndex));
+    return makeString("auto-flow", dense, " ", serializeLonghandValue(autoColumnsIndex), " / ", serializeLonghandValue(columnsIndex));
```
Comment (Loirooriol) [inline Source/WebCore/css/StyleProperties.cpp]: `autoRowsIndex`?

File: Source/WebCore/css/StyleProperties.cpp [hunk 9]
```diff
@@ -829,67 +931,55 @@ String StyleProperties::getLayeredShorthandValue(const StylePropertyShorthand& s
...
 
-    return result.isEmpty() ? String() : result.toString();
+    return result.toString();
 }
 
-String StyleProperties::getGridTemplateValue() const
+String ShorthandSerializer::serializeGridTemplate() const
 {
-    ASSERT(gridTemplateShorthand().length() == 3);
+    ASSERT(length() >= 3);
```
Comment (Loirooriol) [inline Source/WebCore/css/StyleProperties.cpp:968]: I would leave this as an equality, if a new longhand is added in the future, this function will need to be updated.

File: Source/WebCore/css/StyleProperties.cpp [hunk 10]
```diff
@@ -413,179 +547,163 @@ static std::optional<CSSValueID> fontStretchKeyword(double value)
...
-        result.append(result.isEmpty() ? "" : " ", "style"_s);
-    if (caps == CSSValueAuto)
-        result.append(result.isEmpty() ? "" : " ", "small-caps"_s);
-    return result.toString();
+    return makeString(
+        !weight ? "" : "weight",
+        !weight || !style ? "" : " ",
+        !style ? "" : "style",
+        !style || !caps ? "" : " ",
+        !caps ? "" : "small-caps"
```
Comment (Loirooriol) [inline Source/WebCore/css/StyleProperties.cpp]: The separator is missing for `weight = true, style = false, caps = true`.

```suggestion
        weight ? "weight" : "",
        weight && (style || caps) ? " " : "",
        style ? "style" : "",
        style && caps ? " " : "",
        caps ? "small-caps" : ""
```

**Review by Loirooriol — COMMENTED** [review]
Very nice, I was also thinking that using `getPropertyCSSValue` when `commonShorthandChecks` already had the values was not great.

File: Source/WebCore/css/StyleProperties.cpp (hunk 1 above)
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp:348]: Good suggestions. Will do. I was actually thinking about exactly this while considering what to do next and how to adapt this for computed style. Wasteful to look for a shorthand property ID in the property array.

File: Source/WebCore/css/parser/CSSPropertyParser.cpp (hunk 2 above)
Comment (darinadler) [reply Source/WebCore/css/parser/CSSPropertyParser.cpp]: Sure, that will be a bit better.

File: Source/WebCore/css/StyleProperties.cpp (hunk 3 above)
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp]: Got it. I will put in a `FIXME` instead of changing behavior.

File: Source/WebCore/css/StyleProperties.cpp (hunk 4 above)
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp]: OK, I will see how I can improve this.

File: Source/WebCore/css/StyleProperties.cpp (hunk 5 above)
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp]: Yes that's better.

File: Source/WebCore/css/StyleProperties.cpp (hunk 6 above)
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp]: OK

File: Source/WebCore/css/StyleProperties.cpp (hunk 7 above)
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp:1081]: I am not sure those assertions are so great. I agree that they will start failing if we add another longhand to these shorthands but I am not sure a failing assertion is all that helpful. The incorrect serialization seems like the best way to notice the code is incorrect.

But it's easy for me to leave them in.

File: Source/WebCore/css/StyleProperties.cpp (hunk 8 above)
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp]: Oops, I need to find out why tests didn't catch this mistake.

File: Source/WebCore/css/StyleProperties.cpp (hunk 9 above)
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp:968]: This function is shared by a longer shorthand. So I can't change the assertion. Maybe I need to change the function name or add a comment.

File: Source/WebCore/css/StyleProperties.cpp (hunk 10 above)
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp]: Yes, definitely a mistake and I'll do something like that. I am also considering just using a table of the 7 serializations since there are only three Booleans here.

File: Source/WebCore/css/StyleProperties.cpp (hunk 8 above)
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp]: I see now. They did. The relevant test just wasn’t in my short list of tests that I am running locally.

File: Source/WebCore/css/StyleProperties.cpp (hunk 7 above)
Comment (Loirooriol) [reply Source/WebCore/css/StyleProperties.cpp:1081]: My reasoning is that by failing they will attract attention to the the method that needs to have its logic updated. Just checking if the serialization is correct has the problem that not all combinations can be tested, so it may be missed.

Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp:1081]: I think runtime assertions are pretty weak for this purpose, and checking just the length of the shorthand in particular is even weaker, misses things like if we reorder the properties in the shorthand. Instead we should probably switch to static assertions so the serialization code will immediately fail to compile if we change the shorthands. And we should figure out what exactly we are trying to assert. Might be nice to literally repeat the properties in the shorthand right there in the function in an assertion rather than just a comment.

File: Source/WebCore/css/StyleProperties.cpp (hunk 9 above)
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp:968]: See the call from `serializeGrid` below.

Comment (Loirooriol) [reply Source/WebCore/css/StyleProperties.cpp:968]: Maybe keep `gridTemplateShorthand().length() == 3`?
But `length() >= 3` is not much useful, it's already implied by `longhandIndex(2, CSSPropertyGridTemplateAreas)`.

**Review by Loirooriol — APPROVED** [review]

File: Source/WebCore/css/StyleProperties.cpp (hunk 9 above)
Comment (darinadler) [reply Source/WebCore/css/StyleProperties.cpp:968]: OK. I think I will merge what I have and switch back to the `gridTemplateShorthand()` one in a follow up.
